pkg/storage/bucket/user_bucket_client.go

```diff
@@ -12,6 +12,7 @@ import (
 // UserBucketReaderClient is a wrapper around a objstore.BucketReader that reads from user-specific subfolder.
 type UserBucketReaderClient struct {
 	userID string
+	prefix string
```

Comment:
I would suggest to keep prefix-functionality separate from `UserBucketClient`. Once that is done, `UserBucketClient` could be implemented by using `PrefixedBucketClient`. In PR #3811 `UserBucketClient` is getting extra user-specific functionality, but only on write path. WDYT?

That would also allow us to combine multiple prefixed bucket clients together by wrapping them, instead of adding multiple prefix-levels here.

Reply:
I removed any changes to the `UserBucketClient` and implemented the necessary functionality using a newly created `PrefixedBucketClient`.